Feature Swico Syntax S1 / Chores / ESR - QRIID validation

monheimx9 opened the pull request:
Hello bonjour,

This is the new PR from my previous changes to the library.

I'm putting this in production starting Monday to test it further, we're generating around 30 invoices per day, I'll be back soon if something isn't behaving properly

Regarding the formatting of the code
I tried to match your rustfmt settings, saddly with only little success, thus I introduced a rustfmt.toml file directly in the source of this library, feel free to modify it as you wish.

Fix #13

monheimx9 committed:
Initial work on the Swico Sytax Definition

monheimx9 committed:
Rework on the ESR validation and construction

When we're working with the QRR/ESR ref, there's now a check to
determine if the QRIID is maching the ESR ref and vice-versa
Also, the representation of the ESR should always be 27 chars long even
if the reference is actually shorter, paiement processing system can
fail on this (actual case that happen to me)

monheimx9 committed:
Formatting issues and a typo

Address:Cobined was renamed into Address::Combined
Working with a commong rustfmt.toml could be a good start to avoid
unecessary commits
I'm trying to replicate yours but I'm having trouble at this time

monheimx9 committed:
PDF's and chores

Formating

monheimx9 committed:
Making the builder more clearer

Fix hyperlinks issues

jacg commented:
> I tried to match your rustfmt settings

I'll let you into dirty secret: I am a heretic! I believe that authors should make significant effort to make their code easily understandable by humans. One way of doing this involves laying out code in ways that emphasize symmetries and patterns, in order to highlight similarities and key differences and to allow the reader to extract information by merely glancing at the code rather than having to parse it.

Automatic code formatters have a tendency to trample on such carefully-formatted code, and in those cases I eschew them. Regularity is useful, but not more useful than information-rich hand-crafted code. I know, I should be burned at the stake.

TLDR: trying to match my rustfmt settings is futile, because I simply don't allow rustfmt do destroy my careful formatting in significant portions of my code.

One downside of my heresy is that those with more orthodox views are likely to let rustfmt loose on the code, thereby introducing a lot of irrelevant noise into diffs ... and this is what happened in your PR. So I had to spend some time reverting the formatting-only changes, to reveal the functional changes in your PR. Now I have a clean diff to look at when I find some more time to dedicate to this. (Perhaps I should push that version here, so that we have a common point for iterating on any necessary changes, but I'll only do that if and when I get around to asking for any modifications.)

Some of your reformatting changes do regularize code that wasn't initially run through rustfmt for no particularly good reason, and I'd be happy to accept these, but I'd argue that they don't really belong here, certainly not mixed into commits that modify the behaviour of the code.

A few points, after the quickest of glances (not submitting them via GitHub review, as there's a fair chance we might rebase before addressing them):

+ Did you mean to commit `test-example1.pdf`? In principle it should be generated in a temporary directory. In practice I put it in the working directory to make it easier to inspect it when iterating on solutions. I thought I had added it to `.gitignore`, but apparently not.

+ It would be good to replace the magic number `140`, here:

  ```rust
      pub fn add_unstructured(self, text: impl AsRef<str>) -> Result<Self, BillingInfoError> {
          let i = text.as_ref().chars().count();
          if i > 140 {
              return Err(BillingInfoError::Swico(swico::SwicoError::TooLong(i)));
          }
          if let Some(c) = &self.structured() {
              let i = c.chars().count() + i;
              if i > 140 {
                  return Err(BillingInfoError::Swico(swico::SwicoError::TooLong(i)));
              }
          }
  ```
  and even here
  ```rust
  #[error("Maximum 140 characters authorized for BillingInfos, found: {0:?}")]
  ```

+ paiement -> payment

+ `billing_infos/swico/erro.rs` -> `error.rs` (R missing)

monheimx9 commented:
There is beauty in heresy ;)

Well, it means that I really made a mess in the code base, I should have thought about this, please excuse my inexperience, this won't happen again

I would gladly keep your formating instead, I will find a way to automaticaly disable rustfmt under certain conditions, I'm sure Lua and neovim can handle this
Even though this is the first time I experience the rustfmt "getting in my way"
Usually I like what the formatter does, I like regularity and I find the defaults pretty good for my taste, and it's only a matter of taste, and you also have good tastes
If we can find a way to correct this PR I would gladly do it.

Regarding the other points:

- > Did you mean to commit test-example1.pdf?
- It was an overlook from my part, didn't meant to push it


- > It would be good to replace the magic number 140, here:
- Yes, I've replaced some magic numbers but forgot others, I'll correct those as well

- > billing_infos/swico/erro.rs -> error.rs (R missing)
- It was a play on word, on purpose, but I'll change it too :^)

I also spotted a few other typos I made after I pushed the PR, I will also correct those


But before I push any other commit into that PR I will wait for your insight, I don't want to mess it further

jacg commented:
> it means that I really made a mess in the code base, I should have thought about this, please excuse my inexperience, this won't happen again

Please don't think of it that way. (Heresy has its cost, and this is an example. I am more guilty than you.)

>> billing_infos/swico/erro.rs -> error.rs (R missing)

> It was a play on word, on purpose, but I'll change it too :^)

Ah, OK. Unfortunately the play on words is lost within the code itself, where the `r` and `s` are not visible. And people are likely to wonder whether there is a deep reason for the standard name not being used. So it's probably more trouble than its worth.

jacg committed:
Revert all PR #21 changes

jacg committed:
Remove spurious .png from sample data path

jacg committed:
PR #21 without reformatting and spurious PDF

jacg commented:
I've pushed three commits:

1. Revert all your changes. That way we start with a clean slate, but still keep your individual steps for reference.
2. Fix the pre-existing problem with the spurious `.png` in the path in the `--example from-style-guide` tests.
3. All your changes (minus the reformatting, and `test-example1.pdf`) squashed into one commit.

Now we should have a clean diff to make the review easier. We can always rebase it all later, to clean up the story we want to tell, once we're happy with the final product.

Could you check whether it still behaves as you expect?

jacg commented on src/billing_infos/mod.rs:
````diff
@@ -0,0 +1,253 @@
+    /// ```
+    pub fn add_unstructured(self, text: impl AsRef<str>) -> Result<Self, BillingInfoError> {
+        let i = text.as_ref().chars().count();
+        if i > 140 {
````
Replace magic number `140`

jacg commented on src/billing_infos/mod.rs:
```diff
@@ -0,0 +1,253 @@
+    /// let msg = "Invoice F248956-24RI for a new gaming chair";
+    /// let bi = BillingInfos::new().add_unstructured(msg)?;
+    /// assert_eq!(bi.unstructured().unwrap(), msg);
+    /// assert!(bi.structured().is_none());
```
`other_message` belongs to next block, so put the blank line before it.

jacg commented on src/billing_infos/swico/syntax.rs:
```diff
@@ -0,0 +1,169 @@
+
+fn is_date_ok(d: impl AsRef<str>) -> Result<(), Err> {
+    let d = d.as_ref();
+    let lenght = d.chars().count();
```
Typo: lengHT

jacg commented on src/billing_infos/utils.rs:
```diff
@@ -0,0 +1,81 @@
+pub trait Fold {
+    fn fold_from(&self, d: &DataType) -> Option<String>;
+}
+pub trait TotalLenght {
```
Typo: TotalLengHT

jacg commented on src/billing_infos/swico/mod.rs:
```diff
@@ -0,0 +1,147 @@
+
+    #[rstest]
+    fn valid() -> anyhow::Result<()> {
+        let fmt = "%Y-%m-%d";
+        let doc_date = NaiveDate::parse_from_str("2024-06-30", fmt)?;
+        let start_vat = NaiveDate::parse_from_str("2024-05-01", fmt)?;
+        let s = Swico::new()
+            .s1_builder()
+            .add_vat_num("112806097")
+            .add_client_ref(r"145258\/Dépôt")
+            .add_conditions("3:10;0:30")
+            .add_invoice_ref("24073428")
+            .add_vat_date_naive(start_vat, Some(doc_date))
+            .add_doc_date_naive(doc_date)
+            .add_unstructured("73 années de retard d'impôts à payer sous 10 jours")
+            .build()?;
+        let res =
+            String::from(r"//S1/10/24073428/11/240630/20/145258\/Dépôt/30/112806097/31/240501240630/40/3:10;0:30");
+        assert_eq!(s.structured().unwrap(), res);
+        let res = String::from("73 années de retard d'impôts à payer sous 10 jours");
```
Have you considered binding the textual components to variables, and using those as builder arguments and in `res`, rather than typing the same information out twice?

The dates are a pain, because the input and output formats are different, but you've already done half the work there.

monheimx9 replied:
The string from res is an overlook, I wrote these tests as fast and as dirty as I could
Let me rewrite them in a better shape

Regarding the dates issues, I don't mind restricting these functions to only allow NaiveDate types

jacg commented on src/billing_infos/utils.rs:
```diff
@@ -0,0 +1,81 @@
+use crate::billing_infos::{BillingInfoParagrah, DataType};
+use std::collections::HashMap;
+
+pub type RawData = HashMap<DataType, Vec<String>>;
```
Have you considered making `RawData` a separate type, rather than type alias?

There seem to be at least three traits (`Fold`, `TotalLength`, `RawDataKind`) which only seem to be relevant (though maybe I'm failing to understand) for the implementation of this specific functionality on this specific `HashMap`. Would it not be simper, clearer and more robust to make them methods on a genuine `RawData` type? It would also make LSP suggestions more focused.

There's another `HashMap` alias somewhere, where similar comments might apply.

monheimx9 replied:
These traits where made in prevision of a potential future S2 syntax for the Swico spec, as well as other potential specs (like Urbanus ERP, the //U1 syntax)
I was following a similar architecture as we do have this the StructuredAddress and Address types in the main lib
Does it make sense ?

jacg commented on src/billing_infos/swico/builder.rs:
```diff
@@ -0,0 +1,166 @@
+    /// Voucher/Invoice/Bill number
+    ///
+    /// The voucher date is the same as the date of the invoice
+    /// Tt is used as the reference date for the terms and conditions.
```
Typo: Tt -> It

jacg commented on src/billing_infos/swico/mod.rs:
```diff
@@ -0,0 +1,147 @@
+
+const DATE_FMT: &str = "%y%m%d";
+
+type StructuredSet = BTreeMap<SwicoComponent, Arc<str>>;
```
The set of swico components is fixed. Wouldn't it be better to represent `StructuredSet` as a struct with one option field for each possible component? This has a number of advantages:

+ Misspelling a component (key, in the current implementation) is caught at compile time.
+ The values could have more meaningful types (e.g. dates, but you could even create meaningful types with extra validation and helpful methods for other components), rather than having them all be strings.

This ties in to a comment I made elsewhere: I suspect that this type alias could also benefit from being a type in its own right.

monheimx9 replied:
The initial idea for using a BTreeMap with keys and values, was to sort keys in ascending orders (as required by the spec), this way I could add or remove keys without altering the order, and make use of iterators for syntax validation
Let me re-think about it

jacg commented on rustfmt.toml:
```diff
@@ -0,0 +1,8 @@
+unstable_features = true
```
This is not an accurate reflection of how code is formatted in the project, but the presence of a `rustfmt.toml` might be taken to imply that it is. So I would be tempted to leave it out.

But, heretic that I am, I'd rather let @Yatekii decide on this project's approach to automatic formatting.

monheimx9 replied:
I'm mentaly prepared to copy paste #![cfg_attr(rustfmt, rustfmt_skip)] everywhere ;^)

Yatekii replied:
> I'm mentaly prepared to copy paste #![cfg_attr(rustfmt, rustfmt_skip)] everywhere ;^)

big nono :D

`unstable_features = true` is fine :) I generally use stable rustfmt to avoid disagreements, but I prefer some of the unstable styling, so I am ok to use this here :)

jacg commented on src/billing_infos/swico/erro.rs:
```diff
@@ -0,0 +1,9 @@
+#[derive(Debug, thiserror::Error)]
```
As mentioned elsewhere, the play on words in this filename, while amusing, goes against the principle of least surprise, and is probably more trouble than it is worth.

jacg commented on src/utils.rs:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+}
+pub trait IbanKind {
```
What do we gain by making this a trait, rather than simply implementing `kind` as a method on `IbanType`?

monheimx9 replied:
I can't remember why I did it, let's assume we gain nothing
I will change it to a straight impl on the struct

jacg reviewed (commented):
Here's a first set of comments/questions, after a quick glance. There are two main structural themes which recur:

+ You use type aliases for maps, where I suspect separate types might be better.
+ You define a number of traits which are implemented only once, and I'm not seeing where other implementations might arise, so I wonder why the methods aren't implemented directly on the types. (Well, for adding methods to your map aliases, I do see the point, but that need goes away if you replace the aliases with separate types.)

I'd like to hear your opinion on these, before looking any deeper.

I also need to spend some time familiarizing myself with the spec that you are following.

monheimx9 committed:
Namings

Error file renaming

Function renaming

Fixing typos

Fixing magic numbers

Unnecessary trait

monheimx9 committed:
Error handling

monheimx9 committed:
Avoid confusion with date

monheimx9 committed:
Typos

Yatekii commented:
Thanks a lot!

monheimx9 commented:
I'm finally not a virgin anymore <o/
First contribution
Expect better next time, thank you for your patience with me !

Yatekii commented:
All good, super amazing work, thank you :)

jacg commented:
I apologize for not having had the time to finish the review.

Thank you for your contribution!

Yatekii commented:
I finally have more energy & time again to tend to my OSS projects and decided to move this forward :) We can go in small steps and fix things quickly :) `master` can always be bleeding edge and is allowed to break imo :)
